src/Symfony/Component/HttpKernel/Tests/EventListener/LocaleAwareListenerTest.php

```diff
@@ -0,0 +1,131 @@
+    {
+        $this->localeAwareService
+            ->expects($this->at(0))
+            ->method('setLocale')
+            ->will($this->throwException(new \InvalidArgumentException()));
+        $this->localeAwareService
+            ->expects($this->at(1))
+            ->method('setLocale')
+            ->with($this->equalTo('en'));
+
+        $event = new GetResponseEvent($this->createHttpKernel(), $this->createRequest('fr'), HttpKernelInterface::MASTER_REQUEST);
+        $this->listener->onKernelRequest($event);
+    }
+
+    public function testLocaleIsSetInOnKernelFinishRequestWhenParentRequestExists()
+    {
+        $this->localeAwareService
+            ->expects($this->once())
+            ->method('setLocale')
+            ->with($this->equalTo('fr'));
+
+        $this->setMasterRequest($this->createRequest('fr'));
+        $event = new FinishRequestEvent($this->createHttpKernel(), $this->createRequest('de'), HttpKernelInterface::SUB_REQUEST);
+        $this->listener->onKernelFinishRequest($event);
+    }
+
+    public function testLocaleIsNotSetInOnKernelFinishRequestWhenParentRequestDoesNotExist()
+    {
+        $this->localeAwareService
+            ->expects($this->never())
+            ->method('setLocale');
+
+        $event = new FinishRequestEvent($this->createHttpKernel(), $this->createRequest('de'), HttpKernelInterface::SUB_REQUEST);
+        $this->listener->onKernelFinishRequest($event);
+    }
+
+    public function testDefaultLocaleIsUsedOnExceptionsInOnKernelFinishRequest()
+    {
+        $this->localeAwareService
+            ->expects($this->at(0))
+            ->method('setLocale')
+            ->will($this->throwException(new \InvalidArgumentException()));
+        $this->localeAwareService
+            ->expects($this->at(1))
+            ->method('setLocale')
+            ->with($this->equalTo('en'));
+
+        $this->setMasterRequest($this->createRequest('fr'));
+        $event = new FinishRequestEvent($this->createHttpKernel(), $this->createRequest('de'), HttpKernelInterface::SUB_REQUEST);
+        $this->listener->onKernelFinishRequest($event);
+    }
+
+    /**
+     * @expectedException \TypeError
+     */
+    public function testTypeErrorIsThrownWithInvalidServices()
+    {
+        $listener = new LocaleAwareListener(new ArrayIterator(array(new \stdClass())), $this->requestStack);
+
+        $event = new GetResponseEvent($this->createHttpKernel(), $this->createRequest('fr'), HttpKernelInterface::MASTER_REQUEST);
+        $listener->onKernelRequest($event);
+    }
+
+    private function createHttpKernel()
+    {
+        return $this->getMockBuilder('Symfony\Component\HttpKernel\HttpKernelInterface')->getMock();
+    }
+
+    private function createRequest($locale)
+    {
+        $request = new Request();
+        $request->setLocale($locale);
+
+        return $request;
+    }
+
+    private function setMasterRequest($request)
```

Comment:
This method is not necessary when we don't mock the `RequestStack`. It can then be replaced with something like this:

```php
$this->requestStack->push($request);
```

Just make sure to push the subrequests too.